pkg/controller/volume/persistentvolume/pv_controller_base.go

```diff
@@ -368,17 +372,70 @@ func (ctrl *PersistentVolumeController) updateVolumeMigrationAnnotationsAndFinal
 	return newVol, nil
 }
 
-// updateMigrationAnnotationsAndFinalizers takes an Annotations map and finalizers slice and checks for a
-// provisioner name depending if the annotation came from a PVC or not.
-// It will then add a "pv.kubernetes.io/migrated-to" annotation if migration with
-// the CSI driver name for that provisioner is "on" based on feature flags, it will also
-// remove the PV deletion protection finalizer and "pv.kubernetes.io/migrated-to" annotation
-// if migration is "off" for that provisioner in rollback scenarios. Returns true if the annotations
-// map or finalizers were modified and false otherwise.
-// Parameters:
-// - claim: true means the ann came from a PVC, false means the ann came from a PV
-func updateMigrationAnnotationsAndFinalizers(cmpm CSIMigratedPluginManager, translator CSINameTranslator,
-	ann map[string]string, finalizers *[]string, claim bool) bool {
+// modifyDeletionFinalizers updates the finalizers based on the reclaim policy and if it is a in-tree volume or not.
+// The in-tree PV deletion protection finalizer is only added if the reclaimPolicy associated with the PV is `Delete`.
+// The in-tree PV deletion protection finalizer is removed if the reclaimPolicy associated with the PV is `Retain` or
+// `Recycle`, removing the finalizer is necessary to reflect the recalimPolicy updates on the PV.
+// The method also removes any external PV Deletion Protection finalizers added on the PV, this represents CSI migration
+// rollback/disable scenarios.
+func modifyDeletionFinalizers(cmpm CSIMigratedPluginManager, volume *v1.PersistentVolume) ([]string, bool) {
+	modified := false
+	var inFinalizers []string
+	if !utilfeature.DefaultFeatureGate.Enabled(features.HonorPVReclaimPolicy) {
+		return inFinalizers, false
+	}
+	if volume.Finalizers != nil {
+		inFinalizers = append(inFinalizers, volume.Finalizers...)
+	}
+	if !metav1.HasAnnotation(volume.ObjectMeta, storagehelpers.AnnDynamicallyProvisioned) {
+		// PV deletion protection finalizer is currently supported only for dynamically
+		// provisioned volumes.
+		return inFinalizers, false
+	}
+	provisioner := volume.Annotations[storagehelpers.AnnDynamicallyProvisioned]
+	if cmpm.IsMigrationEnabledForPlugin(provisioner) {
+		// Remove in-tree delete finalizer on the PV as migration is enabled.
+		if slice.ContainsString(inFinalizers, storagehelpers.PVDeletionInTreeProtectionFinalizer, nil) {
+			inFinalizers = slice.RemoveString(inFinalizers, storagehelpers.PVDeletionInTreeProtectionFinalizer, nil)
+			modified = true
+		}
+	} else {
+		// Migration disabled.
+		// Check if it is a in-tree volume.
+		if !strings.HasPrefix(provisioner, "kubernetes.io/") {
+			// The provision plugin does not begin with known in-tree plugin volume prefix annotation.
+			return inFinalizers, false
```

Comment:
This can happen also when there is no `AnnDynamicallyProvisioned` annotation. Which is probably correct, but not so obvious.

Reply:
Same as above.